Solution

mralexfed007 opened the pull request:
https://mralexfed007.github.io/js_employees_table_DOM/

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,196 @@
+    body.prepend(...sortedBody);
+    nameSort = !nameSort;
+  } else if (sortTarget.textContent === 'Name') {
+    sortedBody.sort((a, b) => {
+      return b.children[0].textContent.localeCompare(a.children[0].textContent);
+    });
+    body.prepend(...sortedBody);
+    nameSort = !nameSort;
+  }
+
+  if (sortTarget.textContent === 'Position' && positionSort) {
+    sortedBody.sort((a, b) => {
+      return a.children[1].textContent.localeCompare(b.children[1].textContent);
+    });
+    body.prepend(...sortedBody);
+    positionSort = !positionSort;
+  } else if (sortTarget.textContent === 'Position') {
+    sortedBody.sort((a, b) => {
+      return b.children[1].textContent.localeCompare(a.children[1].textContent);
+    });
+    body.prepend(...sortedBody);
+    positionSort = !positionSort;
+  }
+
+  if (sortTarget.textContent === 'Office' && officeSort) {
+    sortedBody.sort((a, b) => {
+      return a.children[2].textContent.localeCompare(b.children[2].textContent);
+    });
+    body.prepend(...sortedBody);
+    officeSort = !officeSort;
+  } else if (sortTarget.textContent === 'Office') {
+    sortedBody.sort((a, b) => {
+      return b.children[2].textContent.localeCompare(a.children[2].textContent);
+    });
+    body.prepend(...sortedBody);
+    officeSort = !officeSort;
+  }
+
+  if (sortTarget.textContent === 'Age' && ageSort) {
+    sortedBody.sort((a, b) => {
```
з сортуванням явно трабли, я писав в попередніх завдання як сортувати правильно.
Не можу це прийняти, код повторюється, виправ попреднє завдання, тоді пофікси тут

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,196 @@
+    ageSort = !ageSort;
+  }
+
+  if (sortTarget.textContent === 'Salary' && salarySort) {
+    sortedBody.sort((a, b) => {
+      const first = Number(a.children[4].textContent
+        .slice(1).split(',').join(''));
+      const second = Number(b.children[4].textContent
+        .slice(1).split(',').join(''));
+
+      return first - second;
+    });
+    body.prepend(...sortedBody);
+    salarySort = !salarySort;
+  } else if (sortTarget.textContent === 'Salary') {
+    sortedBody.sort((a, b) => {
+      const first = Number(a.children[4].textContent
+        .slice(1).split(',').join(''));
+      const second = Number(b.children[4].textContent
+        .slice(1).split(',').join(''));
+
+      return second - first;
+    });
+    body.prepend(...sortedBody);
```
просто повторюється код один в один:(
DRY

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,196 @@
+<label>Position: <input name="position" 
+type="text" data-qa="position" required></label>
+<label>Office: <select name="office" data-qa="office">
+<option>Tokyo</option>
+<option>Singapore</option>
+<option>London</option>
+<option>New York</option>
+<option>Edinburgh</option>
+<option>San Francisco</option>
+</select>
```
Опції пжна було генерувати через map, маючи масив міст де можна працювати

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,196 @@
+const table = document.querySelector('table');
+
+table.insertAdjacentHTML('afterend', `
+<form class="new-employee-form">
+<label>Name: <input name="name" type="text" data-qa="name" required></label>
+<label>Position: <input name="position" 
+type="text" data-qa="position" required></label>
+<label>Office: <select name="office" data-qa="office">
```
чому ми це пишемо тут, немає ні відступів ні нічого, це не читабельно, немає пісвідки синтаксису, шанс зробити помилку в таких умовах рівний 90%

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,196 @@
+</form>
+`);
+
+const form = document.querySelector('form');
```
правильно, спершу створили форму, а тепер її шукаємо)
Навіщо? Можна було об'єкт форми створити через createElement, і працювати з ним

Очікую побачити функцію createForm

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,196 @@
+const form = document.querySelector('form');
+const bodyNode = document.querySelector('body');
+
+form.addEventListener('submit', (e) => {
+  const data = new FormData(form);
+
+  e.preventDefault();
+
+  if (data.get('name').length < 4 || data.get('age') < 18
+  || data.get('age') > 90) {
+    bodyNode.insertAdjacentHTML('afterbegin', `
+    <div class="notification error">
+    <h2 class="title">Error!</h2>
+    <p>Enter valid data, please</p>
+  </div>
+    `);
+
+    setTimeout(() => {
+      bodyNode.children[0].remove();
+    }, 2000);
+  } else {
+    bodyNode.insertAdjacentHTML('afterbegin', `
+    <div class="notification success">
+    <h2 class="title">Success!</h2>
+    <p>Congratulations, new employer has been added!</p>
+  </div>
```
очікую побачити функцію createNotification
не повторюй код!

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,196 @@
+      bodyNode.children[0].remove();
+    }, 2000);
+
+    list.insertAdjacentHTML('beforeend', `
+  <tr>
+    <td>${data.get('name')}</td>
+    <td>${data.get('position')}</td>
+    <td>${data.get('office')}</td>
+    <td>${data.get('age')}</td>
+    <td>$${data.get('salary').match(/.{1,3}/g).join(',')}</td>
+  </tr>
+  `);
+    form.reset();
+  }
```
createTableRow

vpolets reviewed (changes requested):
Виносимо код в блоки які можна перевикористати, створюй допоміжні функції, не треба писати полотна коду який повторюється, потім це дуже важко рефаторити, коли в тебе в багатьох місцях треба щось міняти. Ідеально коли зміна відбувається всього в одному місці, для цього створюй функції і константи

mralexfed007 reviewed (commented):
Не придумал как создать форму более коротко (

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,265 @@
+  });
+
+  switch (item.textContent) {
+    case 'Name':
+      mySort(sortedBody, index, nameSort);
+      nameSort = !nameSort;
+      break;
+
+    case 'Position':
+      mySort(sortedBody, index, positionSort);
+      positionSort = !positionSort;
+      break;
+
+    case 'Office':
+      mySort(sortedBody, index, officeSort);
+      officeSort = !officeSort;
+      break;
+
+    case 'Age':
+      mySort(sortedBody, index, ageSort);
+      ageSort = !ageSort;
+      break;
+
+    case 'Salary':
+      mySort(sortedBody, index, salarySort);
+      salarySort = !salarySort;
+      break;
```
а ці кейси не можна пооб'єднувати?
і не робити купу змінних, одну всього
```
sortedBy = header.textContent
```
щось таке

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,265 @@
+  officeArr.forEach(item => {
+    const selOff = document.createElement('option');
+
+    selOff.textContent = item;
+    selectOffice.append(selOff);
+  });
+  selectOffice.setAttribute('name', 'office');
+  selectOffice.setAttribute('data-qa', 'office');
+  labelOffice.textContent = 'Office: ';
+  labelOffice.append(selectOffice);
+  newForm.append(labelOffice);
+
+  const labelAge = document.createElement('label');
+  const inputAge = document.createElement('input');
+
+  inputAge.setAttribute('name', 'age');
+  inputAge.setAttribute('type', 'number');
+  inputAge.setAttribute('required', true);
+  inputAge.setAttribute('data-qa', 'age');
+  labelAge.textContent = 'Age: ';
+  labelAge.append(inputAge);
+  newForm.append(labelAge);
+
+  const labelSalary = document.createElement('label');
+  const inputSalary = document.createElement('input');
+
+  inputSalary.setAttribute('name', 'salary');
+  inputSalary.setAttribute('type', 'number');
+  inputSalary.setAttribute('required', true);
+  inputSalary.setAttribute('data-qa', 'salary');
+  labelSalary.textContent = 'Salary: ';
+  labelSalary.append(inputSalary);
+  newForm.append(labelSalary);
+
+  const myButton = document.createElement('button');
+
+  myButton.setAttribute('type', 'submit');
+  myButton.textContent = 'Save to table';
+
+  newForm.append(myButton);
```
весь код повторюється, створи собі масив з інпутами і конфігуваціями для них і по цих об'єктах через forEach додай в форму інпутів

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,265 @@
+  e.preventDefault();
+
+  if (data.get('name').length < 4 || data.get('age') < 18
+  || data.get('age') > 90) {
```
```suggestion
    || data.get('age') > 90) {
```

mralexfed007 commented:
а ці кейси не можна пооб'єднувати?
і не робити купу змінних, одну всього

sortedBy = header.textContent


щось таке

при каждом нажатии ведь переменная будет заново перезаписываться и я не
смогу никаких манипуляций с ней сделать. switch я написал, чтобы каждая
колонка помнила какой должен быть текущий порядок сортировки, вне
зависимости от остальных, поэтому и 4 внешних переменных. Оставлю одну
внешнюю и не будет выполнятся условие, что при первом нажатии на
заголовок должна быть сортировка ASC. Или же я просто не понял, как это
сделать

чт, 15 апр. 2021 г. в 12:12, Vladyslav Polets ***@***.***>:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/scripts/main.js
> <https://github.com/mate-academy/js_employees_table_DOM/pull/69#discussion_r613898258>
> :
>
> > +    case 'Name':
>
> +      mySort(sortedBody, index, nameSort);
>
> +      nameSort = !nameSort;
>
> +      break;
>
> +
>
> +    case 'Position':
>
> +      mySort(sortedBody, index, positionSort);
>
> +      positionSort = !positionSort;
>
> +      break;
>
> +
>
> +    case 'Office':
>
> +      mySort(sortedBody, index, officeSort);
>
> +      officeSort = !officeSort;
>
> +      break;
>
> +
>
> +    case 'Age':
>
> +      mySort(sortedBody, index, ageSort);
>
> +      ageSort = !ageSort;
>
> +      break;
>
> +
>
> +    case 'Salary':
>
> +      mySort(sortedBody, index, salarySort);
>
> +      salarySort = !salarySort;
>
> +      break;
>
>
> а ці кейси не можна пооб'єднувати?
> і не робити купу змінних, одну всього
>
> sortedBy = header.textContent
>
>
> щось таке
> ------------------------------
>
> In src/scripts/main.js
> <https://github.com/mate-academy/js_employees_table_DOM/pull/69#discussion_r613898923>
> :
>
> > +  const labelName = document.createElement('label');
>
> +  const inputName = document.createElement('input');
>
> +
>
> +  inputName.setAttribute('name', 'name');
>
> +  inputName.setAttribute('type', 'text');
>
> +  inputName.setAttribute('required', true);
>
> +  inputName.setAttribute('data-qa', 'name');
>
> +  labelName.textContent = 'Name: ';
>
> +  labelName.append(inputName);
>
> +  newForm.append(labelName);
>
> +
>
> +  const labelPos = document.createElement('label');
>
> +  const inputPos = document.createElement('input');
>
> +
>
> +  inputPos.setAttribute('name', 'position');
>
> +  inputPos.setAttribute('type', 'text');
>
> +  inputPos.setAttribute('required', true);
>
> +  inputPos.setAttribute('data-qa', 'position');
>
> +  labelPos.textContent = 'Position: ';
>
> +  labelPos.append(inputPos);
>
> +  newForm.append(labelPos);
>
> +
>
> +  const labelOffice = document.createElement('label');
>
> +  const selectOffice = document.createElement('select');
>
> +
>
> +  officeArr.forEach(item => {
>
> +    const selOff = document.createElement('option');
>
> +
>
> +    selOff.textContent = item;
>
> +    selectOffice.append(selOff);
>
> +  });
>
> +  selectOffice.setAttribute('name', 'office');
>
> +  selectOffice.setAttribute('data-qa', 'office');
>
> +  labelOffice.textContent = 'Office: ';
>
> +  labelOffice.append(selectOffice);
>
> +  newForm.append(labelOffice);
>
> +
>
> +  const labelAge = document.createElement('label');
>
> +  const inputAge = document.createElement('input');
>
> +
>
> +  inputAge.setAttribute('name', 'age');
>
> +  inputAge.setAttribute('type', 'number');
>
> +  inputAge.setAttribute('required', true);
>
> +  inputAge.setAttribute('data-qa', 'age');
>
> +  labelAge.textContent = 'Age: ';
>
> +  labelAge.append(inputAge);
>
> +  newForm.append(labelAge);
>
> +
>
> +  const labelSalary = document.createElement('label');
>
> +  const inputSalary = document.createElement('input');
>
> +
>
> +  inputSalary.setAttribute('name', 'salary');
>
> +  inputSalary.setAttribute('type', 'number');
>
> +  inputSalary.setAttribute('required', true);
>
> +  inputSalary.setAttribute('data-qa', 'salary');
>
> +  labelSalary.textContent = 'Salary: ';
>
> +  labelSalary.append(inputSalary);
>
> +  newForm.append(labelSalary);
>
> +
>
> +  const myButton = document.createElement('button');
>
> +
>
> +  myButton.setAttribute('type', 'submit');
>
> +  myButton.textContent = 'Save to table';
>
> +
>
> +  newForm.append(myButton);
>
>
> весь код повторюється, створи собі масив з інпутами і конфігуваціями для
> них і по цих об'єктах через forEach додай в форму інпутів
> ------------------------------
>
> In src/scripts/main.js
> <https://github.com/mate-academy/js_employees_table_DOM/pull/69#discussion_r613899200>
> :
>
> > +  return tr;
>
> +};
>
> +
>
> +const form = createForm();
>
> +
>
> +table.parentNode.append(form);
>
> +
>
> +const bodyNode = document.querySelector('body');
>
> +
>
> +form.addEventListener('submit', (e) => {
>
> +  const data = new FormData(form);
>
> +
>
> +  e.preventDefault();
>
> +
>
> +  if (data.get('name').length < 4 || data.get('age') < 18
>
> +  || data.get('age') > 90) {
>
>
> ⬇️ Suggested change
>
> -  || data.get('age') > 90) {
>
> +    || data.get('age') > 90) {
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/js_employees_table_DOM/pull/69#pullrequestreview-636455855>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOUVVUIVNE4OS3E4S4TMK4LTI2UYTANCNFSM42ZDPUWQ>
> .
>

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,206 @@
 
-// write code here
+let bySort = null;
+const officeArr = ['Tokyo', 'Singapore', 'London',
```
offices

vpolets commented on src/scripts/main.js:
```diff
@@ -1,3 +1,206 @@
+const table = document.querySelector('table');
+
+const mySort = (arr, index, order) => {
+  if (order.textContent === 'Salary') {
+    if (bySort !== order.textContent) {
+      arr.sort((a, b) => {
+        const first = Number(a.children[index].textContent
+          .slice(1).split(',').join(''));
+        const second = Number(b.children[index].textContent
+          .slice(1).split(',').join(''));
+
+        return first - second;
+      });
+      bySort = order.textContent;
+    } else {
+      arr.reverse();
+    }
+  } else if (order.textContent === 'Age') {
+    if (bySort !== order.textContent) {
+      arr.sort((a, b) => {
+        return a.children[index].textContent - b.children[index].textContent;
+      });
+      bySort = order.textContent;
+    } else {
+      arr.reverse();
+    }
+  } else {
+    if (bySort !== order.textContent) {
+      arr.sort((a, b) => {
+        return a.children[index].textContent
+          .localeCompare(b.children[index].textContent);
+      });
+      bySort = order.textContent;
+    } else {
+      arr.reverse();
```
if if if if hell
є ж switch case було б набагато просторіше, читабельніше і акуратніше

mralexfed007 commented:
скоро переучусь )

пт, 16 апр. 2021 г. в 11:29, Vladyslav Polets ***@***.***>:

> ***@***.**** approved this pull request.
> ------------------------------
>
> In src/scripts/main.js
> <https://github.com/mate-academy/js_employees_table_DOM/pull/69#discussion_r614657332>
> :
>
> > @@ -1,3 +1,206 @@
>
>  'use strict';
>
>
>
> -// write code here
>
> +let bySort = null;
>
> +const officeArr = ['Tokyo', 'Singapore', 'London',
>
>
> offices
> ------------------------------
>
> In src/scripts/main.js
> <https://github.com/mate-academy/js_employees_table_DOM/pull/69#discussion_r614658111>
> :
>
> > +  if (order.textContent === 'Salary') {
>
> +    if (bySort !== order.textContent) {
>
> +      arr.sort((a, b) => {
>
> +        const first = Number(a.children[index].textContent
>
> +          .slice(1).split(',').join(''));
>
> +        const second = Number(b.children[index].textContent
>
> +          .slice(1).split(',').join(''));
>
> +
>
> +        return first - second;
>
> +      });
>
> +      bySort = order.textContent;
>
> +    } else {
>
> +      arr.reverse();
>
> +    }
>
> +  } else if (order.textContent === 'Age') {
>
> +    if (bySort !== order.textContent) {
>
> +      arr.sort((a, b) => {
>
> +        return a.children[index].textContent - b.children[index].textContent;
>
> +      });
>
> +      bySort = order.textContent;
>
> +    } else {
>
> +      arr.reverse();
>
> +    }
>
> +  } else {
>
> +    if (bySort !== order.textContent) {
>
> +      arr.sort((a, b) => {
>
> +        return a.children[index].textContent
>
> +          .localeCompare(b.children[index].textContent);
>
> +      });
>
> +      bySort = order.textContent;
>
> +    } else {
>
> +      arr.reverse();
>
>
> if if if if hell
> є ж switch case було б набагато просторіше, читабельніше і акуратніше
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-academy/js_employees_table_DOM/pull/69#pullrequestreview-637455105>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOUVVULAYYCRGDD2NZKIE63TI7YNLANCNFSM42ZDPUWQ>
> .
>
